connectors/uplynk/src/test/java/com/theoplayer/android/connector/uplynk/internal/UplynkSsaiDescriptionConverterTest.kt

```diff
@@ -0,0 +1,94 @@
+package com.theoplayer.android.connector.uplynk.internal
+
+import com.theoplayer.android.connector.uplynk.UplynkSsaiDescription
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mock
+import org.mockito.MockitoAnnotations
+import org.mockito.kotlin.whenever
+
+class UplynkSsaiDescriptionConverterTest {
+    @Mock
+    private lateinit var ssaiDescription: UplynkSsaiDescription
+
+    private lateinit var converter: UplynkSsaiDescriptionConverter
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
+        whenever(ssaiDescription.prefix).thenReturn("urlprefix")
+        whenever(ssaiDescription.assetIds).thenReturn(listOf("asset1", "asset2", "asset3"))
+        whenever(ssaiDescription.preplayParameters).thenReturn(LinkedHashMap(mapOf("p1" to "v1", "p2" to "v2", "p3" to "v3")))
```

Comment:
Create a real `UplynkSsaiDescription` instead of a mock.

`UplynkSsaiDescription` is already a plain data class. It's easier and more maintainable to use it directly in tests.

Reply:
I tend to use mocks as much as possible because of multiple reasons. Mainly to be sure that I test only one thing. Here it's indeed just a data class, but even here it's constructed using `builder`, which potentially might have some logic that is in my opinion should not fail this test. It's confusing when we have a bug in a `builder` and then `buildPreplayUrl_whenPrefixIsNull_startsUrlFromPrefix` test fails among others. That is not informative.

There is another bonus to use mocks. Look at this test:

```
    @Test
    fun buildPreplayUrl_whenAssetIdHasSingleValue_addsAssetIdAsJsonFilename() {
        whenever(ssaiDescription.assetIds).thenReturn(listOf("singleasset"))

        val result = converter.buildPreplayUrl(ssaiDescription)

        assertTrue(result.contains("/singleasset.json"))
    }
```

Here it tests that we use build urls differently when one property is changed. And from line 1 of the method you directly see which property differs from the basis `ssaiDescription`. In my opinion it gives better view on what we test here than when we use builder:

```
    @Test
    fun buildPreplayUrl_whenAssetIdHasSingleValue_addsAssetIdAsJsonFilename() {
        val ssaiDescription = UplynkSsaiDescription.Builder()
            .prefix("urlprefix")
            .assetIds(listOf("singleasset"))
            .preplayParameters(LinkedHashMap(mapOf("p1" to "v1", "p2" to "v2", "p3" to "v3")))
            .build()

        val result = converter.buildPreplayUrl(ssaiDescription)

        assertTrue(result.contains("/singleasset.json"))
    }
```

But you are not the first who was triggered on mocking source description)) Jeroen also asked me why is that when I showed him this code. 

So I'd not argue here and **### I'll change these tests to use real object** in favour of established practice, but let's consider this discussion for later. I believe that using more mocks has some advantages.